sources/academy/webscraping/scraping_basics_python/11_scraping_variants.md

```diff
@@ -0,0 +1,418 @@
+  </div>
+</div>
+```
+
+These elements aren't visible to regular visitors. They're there just in case JavaScript fails to work, otherwise they're hidden. This is a great find because it allows us to keep our scraper lightweight.
+
+## Extracting variants
+
+Using our knowledge of Beautiful Soup, we can locate the options and extract the data we need:
+
+```py
+...
+
+listing_url = "https://warehouse-theme-metal.myshopify.com/collections/sales"
+listing_soup = download(listing_url)
+
+data = []
+for product in listing_soup.select(".product-item"):
+    item = parse_product(product, listing_url)
+    product_soup = download(item["url"])
+    vendor = product_soup.select_one(".product-meta__vendor").text.strip()
+
+    if variants := product_soup.select(".product-form__option.no-js option"):
+        for variant in variants:
+            data.append(item | {"variant_name": variant.text.strip()})
+    else:
+        item["variant_name"] = None
+        data.append(item)
+
+...
+```
+
+The CSS selector `.product-form__option.no-js` matches elements with both `product-form__option` and `no-js` classes. Then we're using the [descendant combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Descendant_combinator) to match all `option` elements somewhere inside the `.product-form__option.no-js` wrapper.
+
+Python dictionaries are mutable, so if we assigned the variant with `item["variant_name"] = ...`, we'd always overwrite the values. Instead of saving an item for each variant, we'd end up with the last variant repeated several times. To avoid this, we create a new dictionary for each variant and merge it with the `item` data before adding it to `data`. If we don't find any variants, we add the `item` as is, leaving the `variant_name` key empty.
+
+:::tip Python syntax you might not know
+
+Since Python 3.8, you can use `:=` to simplify checking if an assignment resulted in a non-empty value. It's called an _assignment expression_ or _walrus operator_. You can learn more about it in the [docs](https://docs.python.org/3/reference/expressions.html#assignment-expressions) or in the [proposal document](https://peps.python.org/pep-0572/).
+
+Since Python 3.9, you can use `|` to merge two dictionaries. If the [docs](https://docs.python.org/3/library/stdtypes.html#dict) aren't clear enough, check out the [proposal document](https://peps.python.org/pep-0584/) for more details.
+
+:::
+
+If you run the program, you should see 34 items in total. Some items should have no variant:
+
+<!-- eslint-skip -->
+```json title=products.json
+[
+  ...
+  {
+    "variant_name": null,
+    "title": "Klipsch R-120SW Powerful Detailed Home Speaker - Unit",
+    "min_price": "324.00",
+    "price": "324.00",
+    "url": "https://warehouse-theme-metal.myshopify.com/products/klipsch-r-120sw-powerful-detailed-home-speaker-set-of-1",
+    "vendor": "Klipsch"
+  },
+  ...
+]
+```
+
+Some products where we're missing the actual price should now have several variants:
```

Comment:
I actually didn't understand why the price was missing and thought it was a bug until I read to the section where you add the prices.

```suggestion
Some products where we're missing the actual price will now have several variants. We will scrape the price later for those products.
```

Reply:
That's a good catch as well. I'll re-read the lesson and try to make this part more comprehensible.

Reply:
Addressed in [b82b673](https://github.com/apify/apify-docs/pull/1244/commits/b82b673a31d8a7769e5b52c99a94a50b4480029a), thanks!